asyncpg/connection.py

```diff
@@ -1356,15 +1359,27 @@ def _drop_global_statement_cache(self):
             # and https://github.com/MagicStack/asyncpg/issues/76
             # for discussion.
             #
+            await self._maybe_close_bad_connection()
             self._drop_global_statement_cache()
             if self._protocol.is_in_transaction() or not retry:
                 raise
             else:
                 return await self._do_execute(
                     query, executor, timeout, retry=False)
+        except:
+            logging.warning('exception - maybe closing bad connection')
```

Comment:
I would move this into the `_maybe_close_bad_connection` method, and only log a `logging.warning` when we actually close the connection.